domains/client/domain-executor/src/xdm_verifier.rs

```diff
@@ -85,32 +90,38 @@ where
 /// This is used by the System domain to validate Extrinsics.
 /// Returns either true if the XDM is valid else false.
 /// Returns Error when required calls to fetch header info fails.
-pub(crate) fn verify_xdm_with_primary_chain_client<PClient, SClient, PBlock, SBlock>(
+pub(crate) fn verify_xdm_with_primary_chain_client<PClient, PBlock, SBlock, Executor>(
     primary_chain_client: &Arc<PClient>,
-    system_domain_client: &Arc<SClient>,
     extrinsic: &SBlock::Extrinsic,
+    executor: Arc<Executor>,
 ) -> Result<bool, Error>
 where
     PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
     PClient::Api: ExecutorApi<PBlock, SBlock::Hash>,
-    SClient: HeaderBackend<SBlock> + ProvideRuntimeApi<SBlock> + 'static,
-    SClient::Api: MessengerApi<SBlock, NumberFor<SBlock>>,
     SBlock: BlockT,
     PBlock: BlockT,
     NumberFor<PBlock>: From<NumberFor<SBlock>>,
     PBlock::Hash: From<SBlock::Hash>,
+    Executor: CodeExecutor + RuntimeVersionOf,
 {
-    let system_domain_runtime = system_domain_client.runtime_api();
-    let best_hash = system_domain_client.info().best_hash;
-    if let Ok(Some(state_roots)) =
-        system_domain_runtime.extract_xdm_proof_state_roots(best_hash, extrinsic)
-    {
-        // TODO: Can't above variable be reused here?
+    let primary_api = primary_chain_client.runtime_api();
+    let primary_best_hash = primary_chain_client.info().best_hash;
+    let system_domain_wasm = primary_api.system_domain_wasm_bundle(primary_best_hash)?;
+
+    let messenger_api = MessengerApiCompatStateless::<Executor>::with_runtime(
+        system_domain_wasm.to_vec(),
+        executor,
+    );
```

Comment:
I just checked Substrate sources and `self.executor.call()` should cache instances based on hash, but here you have hardcoded static string for hash. If such string is used in other places, we may get two different runtimes with the same hash that will cause hard to debug issues. I think it it worth hashing beforehand.

I really don't like that we're doing heap allocation here with large runtime every time and require hashing of it afterwards too. This should be cached somewhere and reused as much as possible.

Reply:
Agree on the frequent large heap allocation as it will be invoked for every xdm extrinsic on the system domain.

Reply:
The heap allocation could completely avoided except the first time since we are using the executor that automatically cache versioned runtime and if we have a static hash that is reused for this purpose. But with this approach, we will not know runtime upgrades as we continue to use the same cached one. I feel the entire `RuntimeCode` struct fields require heap allocation one way or the other considering the runtime upgrades.

I have a different approach in mind and would need a rebase. Moving this draft until then.